fe/src/main/java/org/apache/doris/load/DeleteHandler.java

```diff
@@ -0,0 +1,627 @@
+     * return false when successfully commit but publish unfinished.
+     * A UserException thrown if both commit and publish failed.
+     * @param job
+     * @param db
+     * @param timeoutMs
+     * @return
+     * @throws UserException
+     */
+    private boolean unprotectedCommitJob(DeleteJob job, Database db, long timeoutMs) throws UserException {
+        long transactionId = job.getTransactionId();
+        GlobalTransactionMgr globalTransactionMgr = Catalog.getCurrentGlobalTransactionMgr();
+        List<TabletCommitInfo> tabletCommitInfos = new ArrayList<TabletCommitInfo>();
+        TabletInvertedIndex invertedIndex = Catalog.getCurrentInvertedIndex();
+        for (TabletDeleteInfo tDeleteInfo : job.getTabletDeleteInfo()) {
+            for (Replica replica : tDeleteInfo.getFinishedReplicas()) {
+                // the inverted index contains rolling up replica
+                Long tabletId = invertedIndex.getTabletIdByReplica(replica.getId());
+                if (tabletId == null) {
+                    LOG.warn("could not find tablet id for replica {}, the tablet maybe dropped", replica);
+                    continue;
+                }
+                tabletCommitInfos.add(new TabletCommitInfo(tabletId, replica.getBackendId()));
+            }
+        }
+        return globalTransactionMgr.commitAndPublishTransaction(db, transactionId, tabletCommitInfos, timeoutMs);
+    }
+
+    /**
+     * This method should always be called in the end of the delete process to clean the job.
+     * Better put it in finally block.
+     * @param job
+     */
+    private void clearJob(DeleteJob job) {
+        if (job != null) {
+            long signature = job.getTransactionId();
+            if (idToDeleteJob.containsKey(signature)) {
+                idToDeleteJob.remove(signature);
+            }
+            for (PushTask pushTask : job.getPushTasks()) {
+                AgentTaskQueue.removePushTask(pushTask.getBackendId(), pushTask.getSignature(),
+                        pushTask.getVersion(), pushTask.getVersionHash(),
+                        pushTask.getPushType(), pushTask.getTaskType());
+            }
+            Catalog.getCurrentGlobalTransactionMgr().getCallbackFactory().removeCallback(job.getId());
+        }
+    }
+
+    public void recordFinishedJob(DeleteJob job) {
+        if (job != null) {
+            long dbId = job.getDeleteInfo().getDbId();
+            LOG.info("record finished deleteJob, transactionId {}, dbId {}",
+                    job.getTransactionId(), dbId);
+            List<DeleteInfo> deleteInfoList = dbToDeleteInfos.get(dbId);
+            if (deleteInfoList == null) {
+                deleteInfoList = Lists.newArrayList();
+                dbToDeleteInfos.put(dbId, deleteInfoList);
+            }
+            deleteInfoList.add(job.getDeleteInfo());
+        }
+    }
+
+    public boolean cancelJob(DeleteJob job, CancelType cancelType, String reason) {
```

Comment:
This method return boolean, but you never use it.
I think return true means cancel succeed(txn failed), and return false means cancel failed(txn succeed).
And the caller should use this return value to decide whether to return user success or failure.

Reply:
And if transaction is COMMITTED but not VISIBLE, you should return a transaction id to user, so that user can use that it to check transaction's state.